qiskit/transpiler/target.py

```diff
@@ -0,0 +1,652 @@
+
+        If the input instruction schedule map contains new instructions not in
+        the target they will be added. However if it contains additional qargs
+        for an existing instruction in the target it will error.
+
+        Args:
+            inst_map (InstructionScheduleMap): The instruction
+            inst_name_map (dict): An optional dictionary that maps any
+                instruction name in ``inst_map`` to an instruction object
+            error_dict (dict): A dictionary of errors of the form::
+
+                {gate_name: {qarg: error}}
+
+            for example::
+
+                {'rx': {(0, ): 1.4e-4, (1, ): 1.2e-4}}
+
+            For each entry in the ``inst_map`` if ``error_dict`` is defined
+            a when updating the ``Target`` the error value will be pulled from
+            this dictionary. If one is not found in ``error_dict`` then
+            ``None`` will be used.
+
+        Raises:
+            ValueError: If ``inst_map`` contains new instructions and
+                ``inst_name_map`` isn't specified
+            KeyError: If a ``inst_map`` contains a qarg for an instruction
+                that's not in the target
+        """
+        for inst in inst_map.instructions:
+            out_props = {}
+            for qarg in inst_map.qubits_with_instruction(inst):
+                sched = inst_map.get(inst, qarg)
+                val = InstructionProperties(calibration=sched)
+                try:
+                    qarg = tuple(qarg)
+                except TypeError:
+                    qarg = (qarg,)
+                if inst in self._gate_map:
+                    if self.dt is not None:
+                        val.duration = sched.duration * self.dt
+                    else:
+                        val.duration = None
+                    if error_dict is not None:
+                        error_inst = error_dict.get(inst)
+                        if error_inst:
+                            error = error_inst.get(qarg)
+                            val.error = error
+                        else:
+                            val.error = None
+                    else:
+                        val.error = None
+                out_props[qarg] = val
+            if inst not in self._gate_map:
+                if inst_name_map is not None:
+                    self.add_instruction(inst_name_map[inst], out_props, name=inst)
+                else:
+                    raise ValueError(
+                        "An inst_name_map kwarg must be specified to add new "
+                        "instructions from an InstructionScheduleMap"
+                    )
+            else:
+                for qarg, prop in out_props.items():
+                    self.update_instruction_properties(inst, qarg, prop)
+
+    @property
+    def qargs(self):
+        """The set of qargs in the target."""
+        if None in self._qarg_gate_map:
+            return None
+        return set(self._qarg_gate_map)
```

Comment:
Likely doesn't matter, but `self._qarg_gate_map.keys()` implements `collections.abc.Set`, so it'll type-check correctly as an immutable set through that API, and it's free to construct.  Both are perhaps better since this is a `property` method; it means it only does trivial work, and you get errors if a user tries to do something like `Target.qargs.add((0,))` to mutate the underlying target.

Reply:
This is a good idea I like this as a return more I'll update this and `operation_names` too